cpu/stm32/periph/fdcan.c

```diff
@@ -0,0 +1,1737 @@
+
+    DEBUG("%s: FDCAN%u GPIO RX = %lu\n",
+          __func__, get_channel_id(can), rx_pin);
+    DEBUG("%s: FDCAN%u GPIO TX = %lu\n",
+          __func__, get_channel_id(can), tx_pin);
+
+    dev->af = af;
+    gpio_init_af(rx_pin, af);
+    gpio_init_af(tx_pin, af);
+}
+
+static int _init(candev_t *candev)
+{
+    can_t *dev = container_of(candev, can_t, candev);
+    FDCAN_GlobalTypeDef *can = dev->conf->can;
+
+    _can[get_channel(can)] = dev;
+
+    /* Erase TX mailbox and RX FIFO */
+    memset(dev->tx_mailbox, 0, sizeof(dev->tx_mailbox));
+    memset(&dev->rx_mailbox, 0, sizeof(dev->rx_mailbox));
+
+    dev->isr_flags.isr_tx = 0;
+    dev->isr_flags.isr_rx = 0;
+
+    /* Enable device clock */
+    periph_clk_en(APB1, dev->conf->rcc_mask);
+
+    DEBUG("%s: FDCAN%u RCC->CCIPR = %lx\n",
+          __func__, get_channel_id(can), RCC->CCIPR);
+
+    _status[get_channel(can)] = STATUS_ON;
+
+    /* configure pins */
+    candev_stm32_set_pins(dev, dev->conf->tx_pin, dev->conf->rx_pin, dev->conf->af);
+
+    set_mode(can, MODE_INIT);
+
+    set_bit_timing(dev);
+
+    /* Enable interrupts */
+    NVIC_EnableIRQ(dev->conf->it0_irqn);
+    NVIC_EnableIRQ(dev->conf->it1_irqn);
+
+    /* Enable auto retransmission on failure*/
+    can->CCCR &= ~FDCAN_CCCR_DAR;
+    /* Enable FD mode */
+    can->CCCR |= FDCAN_CCCR_FDOE;
+    /* Enable bitrate switching */
+    can->CCCR |= FDCAN_CCCR_BRSE;
+    /* Enable transmit pause */
+    can->CCCR |= FDCAN_CCCR_TXP;
+    /* Enable protocol exception handling */
+    can->CCCR &= ~FDCAN_CCCR_PXHD;
+
+#ifdef STM32_PM_STOP
+    pm_block(STM32_PM_STOP);
+#endif
+
+    /* Clear interrupt flags */
+    can->IR |= 0xFFFFFFU;
+    /* Enable all interrupts and both interrupt lines */
+    can->IE |= 0xFFFFFFU;
+
+    can->TXBTIE = FDCAN_TXBTIE_TIE;
+
+    can->ILS = FDCAN_ILS_PERR | FDCAN_ILS_SMSG;
+    can->ILE |= FDCAN_ILE_EINT0 | FDCAN_ILE_EINT1;
+
+    DEBUG("%s: FDCAN%u->CCCR =  %lx\n",
+          __func__, get_channel_id(can), can->CCCR);
+
+    int res = set_mode(can, MODE_NORMAL);
+
+    DEBUG("%s: FDCAN%u init done, return %d\n",
+          __func__, get_channel_id(can), res);
+
+    return res;
+}
+
+static inline void set_bit_timing(can_t *dev)
+{
+    FDCAN_GlobalTypeDef *can = dev->conf->can;
+
+    DEBUG("%s: FDCAN%u setup bittiming\n",
+          __func__, get_channel_id(can));
+    /* Normal CAN bittiming */
+    /* SJW */
+    can->NBTP &= ~FDCAN_NBTP_NSJW;
+    can->NBTP |= (dev->candev.bittiming.sjw - 1) << FDCAN_NBTP_NSJW_Pos;
+    /* Phase Seg 2 */
+    can->NBTP &= ~FDCAN_NBTP_NTSEG2;
+    can->NBTP |= (dev->candev.bittiming.phase_seg2 - 1) << FDCAN_NBTP_NTSEG2_Pos;
+    /* Phase Seg 1 */
+    can->NBTP &= ~FDCAN_NBTP_NTSEG1;
+    can->NBTP |= (dev->candev.bittiming.phase_seg1
+                 + dev->candev.bittiming.prop_seg - 1) << FDCAN_NBTP_NTSEG1_Pos;
+    /* BRP */
+    can->NBTP &= ~FDCAN_NBTP_NBRP;
+    can->NBTP |= (dev->candev.bittiming.brp - 1) << FDCAN_NBTP_NBRP_Pos;
```

Comment:
multiple writes can be avoided here

and for the other registers as well

Reply:
done